### PR TITLE
Remove pending urbr's of a vhci aborted pipe

### PR DESCRIPTION
**Pull request by cezanne**
This pull request is created to discuss several issues from #90.

- IRP cancellation and urbr removal for an aborted pipe. Are they required?
- Verifying a resetting pipe routine(store_urb_reset_pipe)

### CODE REVIEW

**Comment by koudis** [comment]
I will test branch against USB 3.1 drive because without properly "RESET_PIPE" the drive might fail or be really slow so it's easy to find typo bugs. (I will do it thru next weekend)

**Comment by koudis** [comment]
Sorry for delay. We had some problems at work. I will test branch asap.

File: driver/vhci/vhci_read.c [hunk 1]
```diff
@@ -79,30 +82,39 @@ store_urb_reset_dev(PIRP irp, struct urb_req *urbr)
...
 	}
 
-	in = PIPE2DIRECT(urb_rp->PipeHandle);
-	type = PIPE2TYPE(urb_rp->PipeHandle);
-	if (type != USB_ENDPOINT_TYPE_BULK && type != USB_ENDPOINT_TYPE_INTERRUPT) {
-		DBGE(DBG_READ, "Error, not a bulk pipe\n");
-		return STATUS_INVALID_PARAMETER;
-	}
+	/*
+	 * Documentation for USB says:
```
Comment (cezanne) [inline driver/vhci/vhci_read.c]: @koudis :  Check [MSDN comment](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/usb/ns-usb-_urb_header) for URB_FUNCTION_SYNC_RESET_PIPE_AND_CLEAR_STALL
``` 
First, for all pipes except isochronous pipes, this URB sends a CLEAR_FEATURE request to clear the device's ENDPOINT_HALT feature.
```

It seems that URB_FUNCTION_SYNC_RESET_PIPE_AND_CLEAR_STALL is allowed for an iso pipe. 

Type checking code you had removed was a legacy code. Maybe, an old programmer seemed to have a problem that this routine was called on a non-bulk pipe. Also, I do not believe that all drivers or user-level codes behave as usbip-win expects.
I'll revert to a previous revision but improves the code from your work.

Comment (koudis) [reply driver/vhci/vhci_read.c]: So for all pipes except iso  URB URB_FUNCTION_SYNC_RESET_PIPE_AND_CLEAR_STALL is valid?
It's possible. I will test it if it works as expected.

### Note
But due to USB doct there are 4 pipe types

- CONTROL
- ISOCHRONOUS
- BULK
- INTERRUPT

For CONTROL is not necessary to clear STALL explicitly.
For ISO URB URB_FUNCTION_SYNC_RESET_PIPE_AND_CLEAR_STALL wil not be send (doue to Msdn doc).

So I think that your assuptions are right. Thanks!

Comment (cezanne) [reply driver/vhci/vhci_read.c]: @koudis : I think that upper drivers on vhci may call URB_FUNCTION_SYNC_RESET_PIPE_AND_CLEAR_STALL with any type. So the type checking is needed. Please check 35509ef47d56992e4a6ddde178e680b3fc3daa5e .

**Comment by koudis** [comment]
@cezanne I test the branch and it did not work as expected. I Create #106 which repair small typo bugs.

I forget to commit part which add direction to wIndex value for CLEAR_FEATURE request and delete in and PipeHandle argument at setup_csp function (we want to send CLEAR_FEATURE to default control pipe).

Now It works well :).

File: driver/vhci/vhci_read.c (hunk 1 above)
Comment (koudis) [reply driver/vhci/vhci_read.c]: Works after #106

**Comment by cezanne** [comment]
@koudis : Please check e52d8f0a498990ce51ecc9c3cd73f8e27655cfd2. If you confirm, I'll merge into master.

**Comment by koudis** [comment]
@cezanne It works now. With all my USB 3.x drives :). You can merge.

I am just curious about `in` in set_cmd_submit_usbip_header. Can you look at my comment in #106 please? :). Because i am not sure why is needed if is obtained from PipeHandle.

**Comment by cezanne** [comment]
@koudis : Re-work from your comment. 😄 Please, check 48f3818.

**Comment by koudis** [comment]
@cezanne It does nmot work :( (how you can se these 'img' emoji?).

The problem is that you changed (in << 7) to (PIPE2ADDR | PIPE2DIRECT). Pipe to direct returns one of {0, 1}. But we need set last bit of LowByte of wIndex to 1 if in != 0 or 0 if in == 0.

I suggest to retake (in << 7) :D.

**Comment by koudis** [comment]
I will push it to abort_pipe if you agree.

**Comment by koudis** [comment]
@cezanne It works! You can merge now. I have no other questions :blush:

**Comment by cezanne** [comment]
@koudis : My fault! A new f8e1c4c had been uploaded. type colon-type "cry"-choose image will  produce 😢

**Comment by koudis** [comment]
@cezanne thanks for a tip :). I already test it. Works without problem :)

**Comment by cezanne** [comment]
Thanks @koudis . Merged into the master. I'll move on to transfer_ex.

**Pull request closed without merging**